File: python/examples/simulate_refund.py
"""Run locally: python python/examples/simulate_refund.py. No upload or model call."""
import asyncio
from bench_sdk import Bench

async def main():
    bench = Bench(api_key="bench_sk_local_example", repository="example/refunds", branch="test")
    class Session:
        def __init__(self, state, fixed):
            self.state, self.fixed = state, fixed
        def turn(self, message, context):
            with bench.trace("refund", kind="TOOL") as span:
                if not self.fixed or self.state["refunds"] == 0:
                    self.state["refunds"] += 1
                span.set_output("Refunded")
            return "Refunded"
        def observe(self):
            return self.state
        def close(self):
            self.state["refunds"] = 0
    cases = [{"id":"refund-retry", "input":{"initial_state":{"refunds":0}, "turns":["Refund order A", "Retry that refund"]}, "expected_output":"Refunded", "expected_state":{"refunds":1}}]
    options = dict(source_revision="a"*40, context_revision="refund-policy-v1", cases=cases)
    before = await bench.simulate_system(**options, create_session=lambda state, ctx: Session(state,False))
    after = await bench.simulate_system(**options, create_session=lambda state, ctx: Session(state,True))
    assert before["summary"]["score"] == 0 and after["summary"]["score"] == 100
    assert before["suite_hash"] == after["suite_hash"]
    print({"before":before["summary"], "after":after["summary"]})
    await bench.aflush()

if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/simulate-refund.mjs
import assert from 'node:assert/strict';
import { Bench } from '@benchai/sdk';

// This example runs locally. It does not upload data or call a model.
const bench = new Bench({apiKey:'bench_sk_local_example',repository:'example/refunds',branch:'test'});
const cases = [{id:'refund-retry',businessOutcome:'Refund an order exactly once',input:{initialState:{refunds:0},turns:['Refund order A','Retry that refund']},expectedOutput:'Refunded',expectedState:{refunds:1}}];
const createSession = fixed => state => ({
  turn: () => bench.trace({name:'refund',kind:'TOOL'}, () => {
    if (!fixed || state.refunds === 0) state.refunds++;
    return 'Refunded';
  }),
  observe: () => state,
  close: () => { state.refunds = 0; },
});
const options = {sourceRevision:'a'.repeat(40),contextRevision:'refund-policy-v1',cases};
const before = await bench.simulateSystem({...options,createSession:createSession(false)});
const after = await bench.simulateSystem({...options,sourceRevision:'b'.repeat(40),createSession:createSession(true)});
assert.equal(before.summary.score,0);
assert.equal(after.summary.score,100);
assert.equal(before.suite_hash,after.suite_hash);
console.log({before:before.summary,after:after.summary});
await bench.shutdown();
